vmdb/app/models/ext_management_system.rb

```diff
@@ -1,20 +1,26 @@
 class ExtManagementSystem < ActiveRecord::Base
+  SUBCLASSES = %w(
+    EmsInfra
+    EmsCloud
+    EmsContainer
+  )
+
   include EmsRefresh::Manager
 
   def self.types
-    [EmsInfra, EmsCloud].collect(&:types).flatten
+    leaf_subclasses.collect(&:ems_type)
   end
 
   def self.supported_types
-    [EmsInfra, EmsCloud].collect(&:supported_types).flatten
+    supported_subclasses.collect(&:ems_type)
   end
 
   def self.leaf_subclasses
-    [EmsInfra, EmsCloud].collect(&:subclasses).flatten
+    descendants.select { |d| d.subclasses.empty? }
   end
 
   def self.supported_subclasses
-    [EmsInfra, EmsCloud].collect(&:supported_subclasses).flatten
+    subclasses.flat_map(&:supported_subclasses)
```

Comment:
So, i hadn't seen this code until now and I have a hard time groking what leaf_subclasses are vs. supported_subclasses and I don't see tests that explain them in this diff.  Am I missing something?

Ok, I think I know what's messing up my brain...

I had to go into console to understand
```ruby
irb(main):003:0> ExtManagementSystem.descendants.collect(&:name)
=> ["EmsInfra", "EmsMicrosoft", "EmsOpenstackInfra", "EmsRedhat", "EmsVmware", "EmsCloud", "EmsAmazon", "EmsOpenstack", "EmsContainer", "EmsKubernetes"]
irb(main):004:0> ExtManagementSystem.subclasses.collect(&:name)
=> ["EmsInfra", "EmsCloud", "EmsContainer"]
```

It's confusing to me that `leaf_subclasses` calls `descendants` (which includes subclasses, their subclasses, on and on) while `supporte_subclasses` calls `subclasses` (just the subclasses of this class).

I'd prefer renaming `leaf_subclasses` => `leaf_descendants` since it calls descendants.

What do you think @Fryguy?